[KeyInstr][DwarfDebug] Add is_stmt emission support

OCHyams opened the pull request:
Interpret Key Instructions metadata to determine is_stmt placement.

The lowest rank (highest precedent) instructions in each {InlinedAt, atomGroup}
set are candidates for is_stmt. Only the last instruction in each set in a given
block gets is_stmt. Calls always get is_stmt.

OCHyams committed:
[KeyInstr][DwarfDebug] Add is_stmt emission support

Interpret Key Instructions metadata to determine is_stmt placement.

The lowest rank (highest precedent) instructions in each {InlinedAt, atomGroup}
set are candidates for is_stmt. Only the last instruction in each set in a given
block gets is_stmt. Calls always get is_stmt.

OCHyams commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133495?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#133495** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133495?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133495?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#133494** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133494?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133493** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133493?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133492** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133492?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133491** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133491?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133490** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133490?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133489** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133489?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133488** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133488?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133487** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133487?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133486** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133486?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133485** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133485?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133484** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133484?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133483** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133483?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133482** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133482?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133481** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133481?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133480** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133480?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133479** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133479?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133478** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133478?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133477** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133477?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

jmorse commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2333,6 +2352,170 @@ DwarfDebug::emitInitialLocDirective(const MachineFunction &MF, unsigned CUID) {
+        CandidateInsts.push_back(Buoy);
+
+      } else if (CandidateRank == Rank) {
+        // We've seen other instructions in this group of this rank. Discard
+        // ones we've seen in this block, keep the others, add this one.
+        assert(!CandidateInsts.empty());
+        SmallVector<const MachineInstr *> Insts;
+        Insts.reserve(CandidateInsts.size() + 1);
+        for (auto &PrevInst : CandidateInsts) {
+          if (PrevInst->getParent() != MI.getParent())
+            Insts.push_back(PrevInst);
+          else
+            KeyInstructions.erase(PrevInst);
+        }
+        Insts.push_back(Buoy);
+        CandidateInsts = std::move(Insts);
```
Also: this adds the buoy each time, so presumably if you have three identically-sourced instructions in a row, you get three buoys in the candidate list?

OCHyams replied:
Does the first comment / question still apply to this new version of the code?

> Also: this adds the buoy each time, so presumably if you have three identically-sourced instructions in a row, you get three buoys in the candidate list?

If they've got the same atom number, no, there's just one. If they're identical except the atom number yes there will be multiple. That's by design, because IMO we should convey that there's multiple key instructions at the source position and let the debugger do what it wants with that info. (Happy to alter the design if needed, just saying this is working as intended)

jmorse reviewed (commented):
I get the impression that `GroupCandidates` and `KeyInstructions` are being kept strictly in sync; thus couldn't one instead just load KeyInstructions from GroupCandidates once the scan is complete? This avoids filling up the dense map with tombstones.

Am I right in understanding that the buoy means the "least precedence" instruction will get the is_stmt if the highest precedence appears after it in the contiguous blob? (Seems fine, just making sure I understand).

On the whole, the computation function feels like it could be simpler, but in some intangible way I'm not immediately sure of. (Still reading the tests).

jmorse commented on llvm/test/DebugInfo/KeyInstructions/X86/dwarf-buoy-multi-key.mir:
```diff
@@ -0,0 +1,78 @@
+    $ebx = MOV32ri 2, debug-location !DILocation(line: 2, scope: !5, atomGroup: 1, atomRank: 2)
+    $ebx = MOV32ri 3, debug-location !DILocation(line: 2, scope: !5, atomGroup: 1, atomRank: 1)
+    $eax = MOV32ri 4, debug-location !DILocation(line: 2, scope: !5)
+    $eax = MOV32ri 5, debug-location !DILocation(line: 2, scope: !5, atomGroup: 2, atomRank: 1)
```
My understanding is that this instruction should get an is_stmt in its own entry because it's a separate span of atom group 2, is that right? But doesn't (address 0x19 above?) as far as I can tell.

OCHyams replied:
Fixed this by addressing previous comment, good spot.

jmorse reviewed (commented):
Tests: I'd personally prefer the input source and explanation at the top of the file, although this is a style thing.

My understanding of this code is that within a basic block, it should be possible for there to be two sequences of instructions of equal group and rank that both get a buoy and is_stmt, if they're separated by some other atom group. Perhaps I'm wrong; but if I'm right, it probably wants explicit test coverage.

There's a risk that the code being generated is brittle and frequently needs updating; I don't know what probability of that we find acceptable, let's just see how often it needs updating I guess.

OCHyams committed:
format

OCHyams committed:
nits

OCHyams committed:
defer insertion into KeyInstructions - simplifies code

OCHyams committed:
simplfy by using use remove_if

OCHyams committed:
improve comments

OCHyams committed:
simplify: sink common bits

OCHyams committed:
remove some now useless ctrl-flow, keeping asserts

OCHyams committed:
simplify remaining ctrl-flow

OCHyams committed:
simplify call handling

OCHyams reviewed (commented):
> I get the impression that GroupCandidates and KeyInstructions are being kept strictly in sync; thus couldn't one instead just load KeyInstructions from GroupCandidates once the scan is complete? This avoids filling up the dense map with tombstones.

Good observation. That has enabled some code simplification too.

> Am I right in understanding that the buoy means the "least precedence" instruction will get the is_stmt if the highest precedence appears after it in the contiguous blob? (Seems fine, just making sure I understand).

Yep that's right.

> On the whole, the computation function feels like it could be simpler, but in some intangible way I'm not immediately sure of. (Still reading the tests).

This part of the code changed and grew a few times during prototyping; despite efforts to keep it away from the clean/final version some of that cruft evidentially made it into this patch. Looking at it with fresh eyes I've simplified it a bunch (I've split the clean up into few commits so it's easy to track how things moved about).

OCHyams commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2333,6 +2352,170 @@ DwarfDebug::emitInitialLocDirective(const MachineFunction &MF, unsigned CUID) {
+        assert(MI.getDebugLoc() && "Unexpectedly missing DL");
+
+        // Calls are always key.
+        KeyInstructions.insert(Buoy);
```
Updated the comment to highlight that sometimes MI != Buoy.

OCHyams committed:
Update llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp

review nit, use `Group`,  "Avoids the appearance that this could be different from the one earlier."

Co-authored-by: Jeremy Morse <jeremy.morse@sony.com>

OCHyams committed:
fix line entries not folding together properly

OCHyams commented on llvm/test/DebugInfo/KeyInstructions/X86/dwarf-buoy-multi-key.mir:
```diff
@@ -0,0 +1,78 @@
+# DBG-NEXT: 0x000000000000000a      2      0      0   0             0       0
+# DBG-NEXT: 0x000000000000000f      2      0      0   0             0       0
+# DBG-NEXT: 0x0000000000000014      2      0      0   0             0       0
+# DBG-NEXT: 0x0000000000000019      2      0      0   0             0       0
```
Good spot. Fixed in commit "fix line entries not folding together properly". That also fixes a bug where a line should have been is_stmt but wasn't (accidentally floated is_stmt past a different atom, which was unintended)

OCHyams committed:
shuffle input source comment around in tests

OCHyams reviewed (commented):
Tests: I'd personally prefer the input source and explanation at the top of the file, although this is a style thing.

> Done

> My understanding of this code is that within a basic block, it should be possible for there to be two sequences of instructions of equal group and rank that both get a buoy and is_stmt, if they're separated by some other atom group. Perhaps I'm wrong; but if I'm right, it probably wants explicit test coverage.

That is tested in `dwarf-buoy-multi-key.mir` (but was broken, now fixed).

> There's a risk that the code being generated is brittle and frequently needs updating; I don't know what probability of that we find acceptable, let's just see how often it needs updating I guess.

The tests are not fun to read and were not fun to write, and probably won't be particularly fun to maintain as you have pointed out. Not sure this can be massively improved with the current strategy and I couldn't think of a different way to test it, sadly.

OCHyams commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -2333,6 +2352,170 @@ DwarfDebug::emitInitialLocDirective(const MachineFunction &MF, unsigned CUID) {
+  // New function - reset KeyInstructions.
+  KeyInstructions.clear();
+
+  // The current candidate is_stmt instructions for each source atom.
+  // Map {(InlinedAt, Group): (Rank, Instructions)}.
+  DenseMap<std::pair<DILocation *, uint32_t>,
+           std::pair<uint16_t, SmallVector<const MachineInstr *>>>
+      GroupCandidates;
```
Seems much of a muchness. http://llvm-compile-time-tracker.com/compare.php?from=864910b2187803587013f0c6a8ed57361670dc08&to=4ca6a9fa05d51d62b27a48125f97d92812f0972d&stat=instructions%3Au#. The clang build indicates a preference for small-size of 2 rather than 1 when looking at max rss, but I'm not sure I trust that number as I'm pretty sure the clang self host build is release + thinLTO (no -g).

jmorse replied:
Welp, one or two is better than the default SmallVector calculation of "several"

jmorse commented on llvm/test/DebugInfo/KeyInstructions/X86/dwarf-buoy-multi-key.mir:
```diff
@@ -0,0 +1,75 @@
+# DBG-NEXT: 0x0000000000000019      2      0      0   0             0       0  is_stmt
+# DBG-NEXT: 0x000000000000001e      2      0      0   0             0       0  is_stmt
+# DBG-NEXT: 0x0000000000000023      2      0      0   0             0       0  is_stmt
+# DBG-NEXT: 0x0000000000000029      2      0      0   0             0       0  is_stmt end_sequence
```
Is this setting is_stmt on something that isn't an instruction (after `retq`)? It's end_sequence here, but could this leak out into different functions under different circumstances? Or it's just an off-by-one.

OCHyams replied:
It's "normal" - I think the end_sequence entry probably leaves all the other registers untouched to avoid unnecessary bloat? Replicating the is_stmt end_sequence with simple input without key instructions: https://godbolt.org/z/TsbTM8j8q

jmorse reviewed (approved):
LGTM with an inline question (plus please do apply the smallvector-size)

OCHyams committed:
set smallvector inline-elements size
